File: 438. Find All Anagrams in a String/Solution.py
class Solution:
    def findAnagrams(self, s: str, p: str) -> List[int]:
        m, n = len(p), len(s)
        res = []
        if m > n: return res
        freq = [0 for i in range(26)]
        for i in range(m):
            freq[self.getIdx(p[i])] += 1
            freq[self.getIdx(s[i])] -= 1
        if self.check(freq): res.append(0)
        for i in range(m, n):
            freq[self.getIdx(s[i - m])] += 1
            freq[self.getIdx(s[i])] -= 1
            if self.check(freq): res.append(i - m + 1)
        return res
    
    def getIdx(self, c):
        return ord(c) - ord('a')
    
    def check(self, freq):
        for f in freq:
            if f != 0: return False
        return True

File: 438. Find All Anagrams in a String/Solution.java
class Solution {
    public List<Integer> findAnagrams(String s, String p) {
        int m = p.length(), n = s.length();
        List<Integer> res = new ArrayList<>();
        if (m > n) return res;
        int[] freq = new int[26];
        for (int i = 0; i < m; i++) {
            freq[p.charAt(i) - 'a']++;
            freq[s.charAt(i) - 'a']--;
        }
        if (check(freq)) res.add(0);
        for (int i = m; i < n; i++) {
            freq[s.charAt(i - m) - 'a']++;
            freq[s.charAt(i) - 'a']--;
            if (check(freq)) res.add(i - m + 1);
        }
        return res;
    }
    
    public boolean check(int[] freq) {
        for (int f : freq) {
            if (f != 0) return false;
        }
        return true;
    }
}